Code review

voldern opened the pull request:
Hi,

I've done some changes so that the code confirms to the code standard and added some documentation of the new option.

In addition I've changed it so that it resets the timer when `highWaterMark` has been reached.

Thoughts?

Commit:
Rename timeout to flushTimeout

Commit:
Document flushTimeout option

Commit:
Use `.bind()` instead of `self = this`

Commit:
Update test description to fit naming pattern

Commit:
Remove superflous test and use sinon timers

Commit:
Rename timeoutId to flushTimeoutId

Commit:
Clear timeout when highWaterMark has been reached

Commit:
Document flushTimeout in README

Commit:
Remove superflous function in test

rjferguson21 commented:
Thanks @voldern!  Lots of good feedback.  

I'm currently trying to resolve a bug in my current implementation. Wanted to check with you if have any thoughts on how to address it or if it is a non-issue.

With the introduction of a timeout, you are no longer guaranteed to be inserting a queue of size `highWaterMark`.  We probably need to account for the possibility that the records written will not always equal the queue size. 

The relevant lines: 
https://github.com/rjferguson21/elasticsearch-bulk-index-stream/blob/master/index.js#L129-L130

I am testing with another branch that attempts to take that into account by using the records written length instead of the queue length.  It is still a little sloppy right now, but I believe it captures the essence of the potential solution.
https://github.com/rjferguson21/elasticsearch-bulk-index-stream/blob/flush-cb/index.js#L130-L132

voldern commented:
Hi,

That is a very good point, there is a potential race condition lurking in there. Good catch! :)

I'll adapt your suggestions to the current changeset.

Commit:
Clear queue before giving up control of thread

Possible race condition pointed out by rjferguson21:
https://github.com/rjferguson21/elasticsearch-bulk-index-stream/pull/1#issuecomment-193408031.

Clear queue before giving up control of thread.

rjferguson21 commented:
Awesome.  Thanks for the help again!
